### PR TITLE
add env-check

### LINKED ISSUE #1453: Env check for common issues
- [ ] if other cni config in /etc/cni/net.d 
- [ ] if net.ipv4.tcp_mtu_probing equals 1
- [ ] if /proc/sys/net/ipv4/tcp_tw_recycle equals 0
- [ ] if udp inCsumError  keeps increasing
- [ ] if /etc/resolv.conf has other search domain
- [ ] if udp 6081 reachable
- [ ] if any security software exists

### PR DESCRIPTION
**Pull request by hongzhen-ma**
####
add cmd in plugin to check env 
#### Which issue(s) this PR fixes:
Fixes #1453

### CODE REVIEW

File: dist/images/install.sh [hunk 1]
```diff
@@ -3408,6 +3409,23 @@ reload(){
...
 }
 
+env-check(){
+  set +e
+  echo "************************************************************************"
+  echo "The result will be saved to /var/log/kube-ovn/env-check.log on each node"
+  echo "************************************************************************"
+
+  KUBE_OVN_NS=kube-system
+  podNames=`kubectl get pod -n $KUBE_OVN_NS | grep kube-ovn-cni | awk '{print $1}'`
```
Comment (zhangzujian) [inline dist/images/install.sh:3419]: `--no-headers` should be added to kubectl command.

File: dist/images/install.sh [hunk 2]
```diff
@@ -3408,6 +3409,23 @@ reload(){
...
+  for pod in $podNames
+  do
+    kubectl exec -it -n $KUBE_OVN_NS $pod -- bash /kube-ovn/env-check.sh
+  done
+  echo "*************************************************************************"
+  echo "Finish environment check, check result in /var/log/kube-ovn/env-check.log"
```
Comment (zhangzujian) [inline dist/images/install.sh:3425]: We can use `tee` to output result to stdout as well as log file.

File: dist/images/env-check.sh [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+set +e
+checkresult=/var/log/kube-ovn/env-check.log
+echo "Start environment check" > $checkresult
+echo "1) check cni configuration" >> $checkresult
```
Comment (zhangzujian) [inline dist/images/env-check.sh:6]: Considering the manual execution scenarios, we should log the result by `bash env-check.sh | tee /var/log/kube-ovn/env-check.log`.

File: dist/images/env-check.sh [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+if [ $probe_mtu == 0 ]; then
+  echo "The 'tcp_mtu_probing' config may affect traffic, make sure if /proc/sys/net/ipv4/tcp_mtu_probing should be set to 1" >> $checkresult
+fi
+recycle=`cat /proc/sys/net/ipv4/tcp_tw_recycle`
+if [ $recycle == 1 ]; then
+  echo "The 'tcp_tw_recycle' config affects nodeport service, make sure change /proc/sys/net/ipv4/tcp_tw_recycle to 0" >> $checkresult
+fi
+
+echo "3) check checksum value" >> $checkresult
+netstat -s > /dev/null
```
Comment (zhangzujian) [inline dist/images/env-check.sh:28]: we can use `which netstat` to check whether the command exists.

File: dist/images/env-check.sh [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+if [[ $result > 1 ]]; then
+  echo "Found pid with '*defence*' name, make sure it has no effect on traffics" >> $checkresult
+fi
+result=`ps -ef | grep vmsec | wc -l`
+if [[ $result > 1 ]]; then
+  echo "Found pid with '*vmsec*' name, make sure it has no effect on traffics" >> $checkresult
+fi
+
+echo "6) check geneve 6081 connection" >> $checkresult
+nmap -sU 127.0.0.1 -p 6081 > /dev/null
```
Comment (zhangzujian) [inline dist/images/env-check.sh:73]: same

File: dist/images/env-check.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+if [[ $? != 0 ]]; then
+  echo "The nmap cmd not found, maybe can be installed mannully and exec 'nmap -sU 127.0.0.1 -p 6081' to check port connection" >> $checkresult
+else
+  result=`nmap -sU 127.0.0.1 -p 6081`
+  wait
```
Comment (zhangzujian) [inline dist/images/env-check.sh:78]: What does `wait` do?

File: dist/images/env-check.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+fi
+if [ -e /proc/sys/net/ipv4/tcp_tw_recycle ]; then
+  recycle=`cat /proc/sys/net/ipv4/tcp_tw_recycle`
+  if [ $recycle == 1 ]; then
+    echo "The 'tcp_tw_recycle' config affects nodeport service, make sure change /proc/sys/net/ipv4/tcp_tw_recycle to 0"
+  fi
+fi
+
+echo "3) check checksum value"
+which netstat
```
Comment (zhangzujian) [inline dist/images/env-check.sh:31]: Use `which netstat 2>/dev/null >/dev/null` to suppress the output.

File: dist/images/env-check.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+if [[ $result > 1 ]]; then
+  echo "Found pid with '*defence*' name, make sure it has no effect on traffics"
+fi
+result=`ps -ef | grep vmsec | wc -l`
+if [[ $result > 1 ]]; then
+  echo "Found pid with '*vmsec*' name, make sure it has no effect on traffics"
+fi
+
+echo "6) check geneve 6081 connection"
+which nmap
```
Comment (zhangzujian) [inline dist/images/env-check.sh:76]: same

File: dist/images/install.sh [hunk 9]
```diff
@@ -3408,6 +3409,17 @@ reload(){
...
 }
 
+env-check(){
+  set +e
+
+  KUBE_OVN_NS=kube-system
+  podNames=`kubectl get pod --no-headers -n $KUBE_OVN_NS | grep kube-ovn-cni | awk '{print $1}'`
+  for pod in $podNames
+  do
+    kubectl exec -it -n $KUBE_OVN_NS $pod -c cni-server -- bash /kube-ovn/env-check.sh | tee
```
Comment (zhangzujian) [inline dist/images/install.sh:3419]: missing log file here.

File: dist/images/env-check.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+if [[ $? != 0 ]]; then
+  echo "The nmap cmd not found, maybe can be installed mannully and exec 'nmap -sU 127.0.0.1 -p 6081' to check port connection"
+else
+  result=`nmap -sU 127.0.0.1 -p 6081`
+  wait
+  if [[ ! $result =~ "open" ]]; then
+    echo "The 6081 port for geneve encapsulation may be not available, please check if ovs-ovn pod is health"
```
Comment (zhangzujian) [inline dist/images/env-check.sh:83]: replace `health` with `healthy`

File: dist/images/install.sh (hunk 1 above)
Comment (hongzhen-ma) [reply dist/images/install.sh]: modified

File: dist/images/install.sh (hunk 2 above)
Comment (hongzhen-ma) [reply dist/images/install.sh]: modified.

使用tee输出到终端之后，就不用再单独保存到文件了。

File: dist/images/env-check.sh (hunk 3 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh]: modified

File: dist/images/env-check.sh [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+echo "2) check system ipv4 config" >> $checkresult
+probe_mtu=`cat /proc/sys/net/ipv4/tcp_mtu_probing`
+if [ $probe_mtu == 0 ]; then
+  echo "The 'tcp_mtu_probing' config may affect traffic, make sure if /proc/sys/net/ipv4/tcp_mtu_probing should be set to 1" >> $checkresult
+fi
+recycle=`cat /proc/sys/net/ipv4/tcp_tw_recycle`
+if [ $recycle == 1 ]; then
+  echo "The 'tcp_tw_recycle' config affects nodeport service, make sure change /proc/sys/net/ipv4/tcp_tw_recycle to 0" >> $checkresult
+fi
```
[the remark this answers is not in the document]
Comment (hongzhen-ma) [reply dist/images/env-check.sh]: modified

File: dist/images/env-check.sh (hunk 4 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh]: modified

File: dist/images/env-check.sh (hunk 5 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh]: modified

File: dist/images/env-check.sh (hunk 6 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh:81]: deleted
之前测试的时候，需要等到命令行执行完，拿到全部的输出结果。现在测试删掉wait之后也可以。

File: dist/images/env-check.sh (hunk 7 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh:31]: modified

File: dist/images/install.sh (hunk 9 above)
Comment (hongzhen-ma) [reply dist/images/install.sh:3419]: 使用tee输出到终端之后，就不用再单独写文件了。thanks

File: dist/images/env-check.sh (hunk 10 above)
Comment (hongzhen-ma) [reply dist/images/env-check.sh:83]: modified

File: dist/images/install.sh [hunk 12]
```diff
@@ -3408,6 +3409,20 @@ reload(){
...
 
+env-check(){
+  set +e
+
+  KUBE_OVN_NS=kube-system
+  podNames=`kubectl get pod --no-headers -n $KUBE_OVN_NS | grep kube-ovn-cni | awk '{print $1}'`
+  for pod in $podNames
+  do
+    echo "****************************************************"
+    echo "Start environment check in pod $pod"
```
Comment (zhangzujian) [inline dist/images/install.sh:3420]: Would be better to print `Start environment check for node $node`.

**Comment by zhangzujian** [comment]
One more thing, we need to add `env-check.sh` to the kube-ovn image.

File: dist/images/install.sh (hunk 12 above)
Comment (hongzhen-ma) [reply dist/images/install.sh]: modified

**Review by zhangzujian — APPROVED** [review]
LGTM

**Comment by hongzhen-ma** [comment]
已回合release-1.8 和 release-1.9 分支

**Pull request merged**